sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3315,6 +3329,54 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
     }
   }
 
+  /**
+   * Resolves columns of an output table from the data in a logical plan. This rule will:
+   *
+   * - Insert casts when data types do not match
+   * - Detect plans that are not compatible with the output table and throw AnalysisException
+   */
+  object ResolveSetVariable extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+      _.containsPattern(COMMAND), ruleId) {
+      case setVariable: SetVariable
+        if setVariable.sourceQuery.resolved && !setVariable.targetVariables.forall(_.resolved) =>
+
+        /**
+         * Resolve the left hand side of the SET
+         */
+        val resolvedVars = setVariable.targetVariables.map { variable =>
+            variable match {
+            case v: UnresolvedVariable =>
+              val varIdent = VariableIdentifier(v.nameParts)
+              val varInfo = sessionCatalog.getVariable(varIdent)
+              if (!varInfo.isDefined) {
+                throw  unresolvedVariableError(varIdent, Seq("SESSION"))
+              }
+              VariableReference(varIdent.variableName, varInfo.get._1, canFold = false)
+            case other => other
+          }
+        }
+
+        /**
+         * Protect against duplicate variable names
+         */
+        val varNames = resolvedVars.collect { case variable => variable.toString }
+        val dups = varNames.diff(varNames.distinct).distinct
```

Comment:
do we need to consider case sensitivity here? Leave a brief comment?

Reply:
Done. The name is formatted when we create the variable (same as we do for other objects such as schema names)